device_samples/getting_started/src/connections/connection_string.js

```diff
@@ -0,0 +1,67 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+'use strict';
+
+// Choose a protocol by uncommenting one of these transports.
+const Protocol = require('azure-iot-device-mqtt').Mqtt;
+// const Protocol = require('azure-iot-device-amqp').Amqp;
+// const Protocol = require('azure-iot-device-http').Http;
+// const Protocol = require('azure-iot-device-mqtt').MqttWs;
+// const Protocol = require('azure-iot-device-amqp').AmqpWs;
+
+const Client = require('azure-iot-device').Client;
+const Message = require('azure-iot-device').Message;
+
+// String containing Hostname, Device Id & Device Key in the following formats:
+//  "HostName=<iothub_host_name>;DeviceId=<device_id>;SharedAccessKey=<device_key>"
+const deviceConnectionString = process.env.IOTHUB_DEVICE_CONNECTION_STRING;
+
+// make sure we have a connection string before we can continue
+if (deviceConnectionString === '' || deviceConnectionString === undefined) {
+  console.error('\x1b[31m%s\x1b[0m', 'Missing device connection string');
+  process.exit(0);
+}
+
+function generateMessage() {
+  const data = JSON.stringify({
+    deviceId: 'my-first-device',
+    temperature: 20 + Math.random() * 10,
+  });
+  
+  return new Message(data);
+}
+
+// fromConnectionString must specify a transport constructor, coming from any transport package.
+const client = Client.fromConnectionString(deviceConnectionString, Protocol);
+const message = generateMessage();
+
+// open client connection
+client.open().catch((err) => {
+  console.error('\x1b[31m%s\x1b[0m', 'Could not connect: ' + err.message);
+  process.exit(0);
+});
+console.log('Client connection: Open');
+
+// send a messge
+console.log('Sending message: ' + message.getData());
+client.sendEvent(message, printResultFor('Send'));
```

Comment:
If I'm reading this right, `client.sendEvent` returns a `Promise` object, but this sample never waits for this Promise to complete.  And looking above, shouldn't there be a `.then` after the `client.open().catch()`? I think the `console.log` on line 44 _could_ execute before the client is actually open.  There's so many ways to write in JS now that I'm not even sure anymore.

Reply:
Sorry, I'm wrong, it doesn't return a `Promise` because we're passing in a callback via `printResultsFor`. But, the concern stands that code after this _could_ execute before the message is actually sent.

I personally think this `printResultsFor` is an artifact of a coding shortcut written in ancient times based on the limitations of the day.  This pattern should be stricken from our samples, and we should decide if we want our JS to use:
1. callbacks
2. .Promise chains (then.then.then.catch)
or
3. await

It seems like we're living with a combination of all 3, and I'm guilty of using all 3 together, so I'm not sure if there's a better option, but we should talk about it.

Reply:
Agreed, and I'd vote for async/await-ifying wherever reasonable.